src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala

```diff
@@ -0,0 +1,156 @@
+
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+    def genLitFuncName(lit: String): String = stringBeginning + Hex.encodeHexString(lit.getBytes("UTF-8"))
+
+    default(super.expr(ctx)) {
+      case (e: in.DfltVal) :: ctx.String() / Exclusive =>
+        val litFuncName = genLitFuncName("") // "" is the default string value
+        strLengths += litFuncName -> 0
+        unit(withSrc(vpr.DomainFuncApp(func = makeFunc(litFuncName), Seq(), Map.empty), e))
+      case lit: in.StringLit if lit.typ.addressability == Exclusive =>
+        val litFuncName = genLitFuncName(lit.s)
+        strLengths += litFuncName -> lit.s.length
+        unit(withSrc(vpr.DomainFuncApp(func = makeFunc(litFuncName), Seq(), Map.empty), lit))
+      case len@in.Length(exp :: ctx.String()) =>
+        for { e <- goE(exp) } yield withSrc(vpr.DomainFuncApp(func = lenFunc, Seq(e), Map.empty), len)
+      case concat@in.Concat(l :: ctx.String(), r :: ctx.String()) =>
+        for {
+          lEncoded <- goE(l)
+          rEncoded <- goE(r)
+        } yield withSrc(vpr.DomainFuncApp(concatFunc, Seq(lEncoded, rEncoded), Map.empty),concat)
+    }
+  }
+
+  override def finalize(col: Collector): Unit = {
+    col.addMember(genDomain())
+  }
+
+  /** Every string literal in the program is encoded as a unique domain function in the String domain,
+    * whose value corresponds to the string id.
+    *     unique function stringLitX(): Int
+    * Here, X is an unique suffix depending on the literal value.
+    */
+  private def makeFunc(name: String): vpr.DomainFunc = {
+    val func = vpr.DomainFunc(
+      name = name,
+      formalArgs = Seq(),
+      typ = vpr.Int,
+      unique = true
+    )(domainName = domainName)
+    funcs += name -> func
+    func
+  }
+
+  /**
+    * Generates
+    *   function strLen(id: Int): Int
+    */
+  private val lenFuncName: String = "strLen"
+  private val lenFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = lenFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl("id", vpr.Int)()),
+    typ = vpr.Int,
+  )(domainName = domainName)
+
+  /**
+    * Generates
+    *   function strConcat(l: Int, r: Int): Int
+    * where l and r are string ids
+    */
+  private val concatFuncName: String = "strConcat"
+  private val concatFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = concatFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl("l", vpr.Int)(), vpr.LocalVarDecl("r", vpr.Int)()),
+    typ = vpr.Int,
+  )(domainName = domainName)
+
+  private def genDomain(): vpr.Domain = {
+    funcs += lenFuncName -> lenFunc
+    funcs += concatFuncName -> concatFunc
+
+    /**
+      * The length of every string literal in the program is axiomatized as
+      *   axiom {
+      *     strLen(literal()) == X
+      *   }
+      * where `literal` is one of the generated unique domain functions and X is the length of the corresponding string
+      */
+    val lenAxioms = strLengths.toSeq.map {
+      case (str, len) =>
```

Comment:
You wouldn't need to store `len` in the map as you could calculate it here for the given `str`. However, you do not have to change it as it's so minor

Reply:
Yes, it is indeed wasteful. Fixed that in the last commit